File: library/src/main/java/renetik/android/view/CSPagerController.java
package renetik.android.view;

import android.content.res.Configuration;
import androidx.viewpager.widget.ViewPager;

import renetik.android.view.adapter.CSOnPageChanged;
import renetik.android.view.adapter.CSOnPageDragged;
import renetik.android.viewbase.CSView;
import renetik.android.viewbase.CSViewController;
import renetik.java.collections.CSList;

import static renetik.android.lang.CSLang.NO;
import static renetik.android.lang.CSLang.YES;
import static renetik.android.lang.CSLang.debug;
import static renetik.android.lang.CSLang.doLater;
import static renetik.android.lang.CSLang.empty;
import static renetik.android.lang.CSLang.equal;
import static renetik.android.lang.CSLang.info;
import static renetik.android.lang.CSLang.is;
import static renetik.android.lang.CSLang.iterate;
import static renetik.android.lang.CSLang.set;
import static renetik.android.lang.CSLang.size;

public class CSPagerController<Page extends CSPagerPage> extends CSViewController<ViewPager> {

    private CSList<Page> _controllers;
    private CSView _emptyView;
    private int _tabLayout;
    private Integer _currentIndex;
    private int _index;

    public CSPagerController(CSViewController parent, int pagerId, int tabs) {
        this(parent, pagerId, tabs, null);
    }

    public CSPagerController(CSViewController parent, int pagerId, CSList<Page> controllers) {
        this(parent, pagerId, 0, controllers);
    }

    public CSPagerController(CSViewController parent, int pagerId, int tabLayout, CSList<Page> controllers) {
        super(parent, pagerId);
        _tabLayout = tabLayout;
        _controllers = controllers;
    }

    public CSView setEmptyView(int id) {
        return setEmptyView(new CSView<>(parent(), id));
    }

    public CSView setEmptyView(CSView view) {
        _emptyView = view;
        if (is(_emptyView)) _emptyView.visible(empty(_controllers));
        return _emptyView;
    }

    public CSPagerController reload(CSList<Page> controllers) {
        int currentItem = getView().getCurrentItem();
        for (Page iController : iterate(_controllers)) {
            iController.asController().onDeinitialize(null);
            iController.asController().onDestroy();
        }
        _controllers = controllers;
        updateControllersState(_controllers.length() > currentItem ? currentItem : 0);
        for (Page page : _controllers) page.asController().initialize(getState());
        if (_controllers.length() > currentItem) getView().setCurrentItem(currentItem, YES);
        updateView();
        return this;
    }

    private void updateView() {
        if (is(_controllers)) getView().setAdapter(new CSPagerAdapter(_controllers));
        visible(set(_controllers));
        if (is(_emptyView)) _emptyView.visible(empty(_controllers));
    }

    protected void onCreate() {
        super.onCreate();
        new CSOnPageDragged(this)
                .onDragged(draggedIndex -> _controllers.get(draggedIndex).asController().setShowingInContainer(YES))
                .onReleased(draggedIndex -> {
                    if (!equal(currentIndex(), draggedIndex))
                        _controllers.get(draggedIndex).asController().setShowingInContainer(NO);
                });
        getView().addOnPageChangeListener(new CSOnPageChanged(currentIndex -> {
            debug("OnPageChanged", currentIndex);
            doLater(100, () -> updateControllersState(currentIndex));
        }));
        updateView();
        updateTabsVisibility();
    }

    private void updateTabsVisibility() {
        if (set(_tabLayout)) item(_tabLayout).fade(isPortrait());
    }

    private void updateControllersState(int currentIndex) {
        if (equal(_currentIndex, currentIndex)) return;
        _currentIndex = currentIndex;
        if (empty(_controllers)) return;
        for (int index = 0; index < size(_controllers); index++)
            _controllers.get(index).asController().setShowingInContainer(index == currentIndex ? YES : NO);
    }

    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        updateTabsVisibility();
    }

    public Page currentController() {
        if (empty(_controllers)) return null;
        return _controllers.get(_currentIndex);
    }

    public void currentIndex(int index) {
        getView().setCurrentItem(index);
    }

    public Integer currentIndex() {
        return _currentIndex;
    }

    public CSList<Page> controllers() {
        return _controllers;
    }

    public boolean isLoaded() {
        return is(_controllers);
    }
}

File: library/src/main/java/renetik/android/view/pager/CSPagerController.kt
package renetik.android.view.pager

import android.view.View
import androidx.viewpager.widget.ViewPager
import renetik.android.extensions.view.visible
import renetik.android.java.collections.CSList
import renetik.android.java.collections.list
import renetik.android.lang.CSLang.*
import renetik.android.view.adapter.CSOnPageChange
import renetik.android.viewbase.CSViewController

class CSPagerController<PageType>(parent: CSViewController<*>, pagerId: Int)
    : CSViewController<ViewPager>(parent, pagerId)
        where PageType : CSViewController<*>, PageType : CSPagerPage {

    val controllers = list<PageType>()
    var currentIndex = 0
    private var emptyView: View? = null

    constructor(parent: CSViewController<*>, pagerId: Int, pages: CSList<PageType>)
            : this(parent, pagerId) {
        controllers.append(pages)
    }

    fun emptyView(view: View) = apply { emptyView = view.visible(empty(controllers)) }

    fun reload(pages: CSList<PageType>) = apply {
        val currentIndex = view.currentItem
        for (page in controllers) {
            page.onDeinitialize(null)
            page.onDestroy()
        }
        controllers.reload(pages)
        updatePageVisibility(if (pages.length() > currentIndex) currentIndex else 0)
        for (page in pages) page.initialize(state)
        view.setCurrentItem(currentIndex, YES)
        updateView()
    }

    override fun onCreate() {
        super.onCreate()
        CSOnPagerPageChange(this)
                .onDragged { index -> controllers[index].showingInContainer(YES) }
                .onReleased { index -> if (currentIndex != index) controllers[index].showingInContainer(NO) }
        view.addOnPageChangeListener(
                CSOnPageChange { index -> doLater(100) { updatePageVisibility(index) } })
        updateView()
    }

    private fun updateView() {
        view.adapter = CSPagerAdapter(controllers)
        visible(controllers.hasItems)
        emptyView?.visible(controllers.isEmpty())
    }

    private fun updatePageVisibility(newIndex: Int) {
        if (equal(currentIndex, newIndex)) return
        currentIndex = newIndex
        for (index in 0 until controllers.size)
            controllers[index].showingInContainer(if (index == currentIndex) YES else NO)
    }

    fun currentController() = controllers.at(currentIndex)

    fun setCurrentIndex(index: Int) = apply { view.currentItem = index }
}
